### PR TITLE
✨ Autoscaling: handle application restart and in parallel  (⚠️ devops)

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?

based on https://github.com/ITISFoundation/osparc-simcore/pull/3689 (this needs to be merged beforehand)

This PR defines a decorator that allows for running **exclusively** any python method, even accross processes by leveraging the Redis Locks. 
The ```exclusive``` decorator is used to protect the autoscaling service background task so **only one** instance of autoscaling can run the task at the same time. Therefore this prevents mulitple scaling up or down of the cluster.

 (⚠️ devops):
- REDIS_HOST
- REDIS_PORT

## Related issue/s
solves #3628

### CODE REVIEW

File: packages/dask-task-models-library/requirements/_base.txt [hunk 1]
```diff
@@ -1,6 +1,6 @@
 #
-# This file is autogenerated by pip-compile with Python 3.9
-# by the following command:
+# This file is autogenerated by pip-compile with python 3.9
```
Comment (pcrespov) [inline packages/dask-task-models-library/requirements/_base.txt:2]: i have the impression we are using different pip-tools version. To be sure, just 
```
cd requirements/tools
make build-nc
make shell
# cd to the requirments folder
$ mak reqs
```

**Review by pcrespov — COMMENTED** [review]
is this ready for review??

**Review by pcrespov — APPROVED** [review]
very nice!

File: packages/service-library/src/servicelib/redis_utils.py [hunk 2]
```diff
@@ -1,25 +1,34 @@
...
+    lock_key: a string as the name of the lock (good practice: app_name:lock_name)
+    lock_value: some additional data that can be retrieved by another client
+    """
 
-    WAIT_SECS = 2
-    ATTEMPTS_COUNT = 20
+    def decorator(func):
+        @functools.wraps(func)
+        async def wrapper(*args, **kwargs):
+            async with redis.lock_context(lock_key=lock_key, lock_value=lock_value):
```
Comment (pcrespov) [inline packages/service-library/src/servicelib/redis_utils.py:29]: Q: so this ``lock_context``ensures as well unlocking if the service is restarted, right?

Comment (pcrespov) [reply packages/service-library/src/servicelib/redis_utils.py:29]: Q: if redis has a lock, does it wait or raises?

Comment (sanderegg) [reply packages/service-library/src/servicelib/redis_utils.py:29]: yes, the lock is TTLed

Comment (sanderegg) [reply packages/service-library/src/servicelib/redis_utils.py:29]: > Q: if redis has a lock, does it wait or raises?

Could you precise this? what do you mean by redis has a lock?

**Review by mrnicegyu11 — APPROVED** [review]
thanks!

File: packages/service-library/src/servicelib/redis_utils.py (hunk 2 above)
Comment (GitHK) [reply packages/service-library/src/servicelib/redis_utils.py:29]: I think he meant if the lock in redid is already locked, what happens?

File: services/autoscaling/src/simcore_service_autoscaling/modules/redis.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+def setup(app: FastAPI) -> None:
+    async def on_startup() -> None:
+        app.state.redis = None
+        settings: RedisSettings = app.state.settings.AUTOSCALING_REDIS
+        app.state.redis = client = RedisClientSDK(settings.dsn_locks)
+        async for attempt in AsyncRetrying(
+            reraise=True,
+            stop=stop_after_delay(120),
+            wait=wait_random_exponential(max=30),
```
Comment (GitHK) [inline services/autoscaling/src/simcore_service_autoscaling/modules/redis.py:25]: 👍

**Review by GitHK — APPROVED** [review]
👍 nice

File: packages/service-library/src/servicelib/redis_utils.py (hunk 2 above)
Comment (sanderegg) [reply packages/service-library/src/servicelib/redis_utils.py:29]: ok, it will raise ```AlreadyLockedError```

**Pull request merged**